docs/proposals/policy.md

```diff
@@ -0,0 +1,146 @@
+
+
+Policy extends the existing authorization rules in Kubernetes to fulfill requirements around extended verb sets, verb specific attributes, role based management of attribute based access control, and more efficient evaluation of authorization status at larger scales.
+
+
+### Runtime Authorization Types
+```
+type Authorizer interface{
+	// Authorize can return both allowedBy and deniedBecause strings at the same time.  
+	// This can happen when one rule allows an action, but another denies the action.
+	// Allowing both to be returned makes it easier to track why policy is allowing or
+	// denying a particular action.  Allowed by must be non-empty and deniedBecause must
+	// be empty in order for an action to be allowed.
+	Authorize(a AuthorizationAttributes) (allowedBy string, deniedBecause string, error)
+}
+
+// AuthorizationAttributeBuilder takes a request and creates AuthorizationAttributes.  
+// Since the attributes returned can vary based on verb type, AuthorizationAttributeBuilders 
+// are paired with Authorizers during registration.
+type AuthorizationAttributeBuilder interface{
+	GetAttributes(request http.Request) (AuthorizationAttributes, error)
+}
+
+type AuthorizationAttributes interface{
+	GetUserInfo() UserInfo  // includes GetName() string GetGroups() []string
+	GetVerb() string
+	GetNamespace() string
+	// GetRequestAttributes is of type interface{} because different verbs and different 
+	// Authorizer/AuthorizationAttributeBuilder pairs may have different contract requirements
+	GetRequestAttributes() interface{}
+}
+```
+A single Authorizer/AuthorizationAttributeBuilder pair may be registered with the apiserver.
+
+
+### Verb Set
+Currently, we only have "readonly" and everything else.  This does not give sufficient control.  We need to support all of the standard HTTP verbs, plus different concepts such as "watch" that don't have equivalents.  Our authorizer will start with the following verbs, but other implementations could choose different ones:
+  1.  exec
+  2.  watch
+  3.  list
+  4.  get
+  5.  create
+  6.  update
+  7.  delete
+  8.  proxy
+
+
+### Verb Specific Attributes
+Different verbs will require different attributes.  For instance, the exec verb will have information about the command being run, while a get will have the resource kind being requested, and an update may have a list of the fields being modified.  The exact contract rules will be chosen by the Authorizer/AuthorizationAttributeBuilder pair, but basic resource access for openshift should include  kind.
+
+
+### Configtime Authorization Types
+```
+// PolicyRule holds information that describes a policy rule, but does not contain information 
```

Comment:
This section would be much better showing actual policies, roles, and bindings rather than starting with the types.  Show an example policy that recreates a realistic scenario, then you can say here are the types that materialize it.